### PR TITLE
feat: accept JSON-formatted Shadowsocks Config

### PR DESCRIPTION
**Pull request by jyyi1**
A new function `shadowsocks.NewClientFromJSON(string)` is added, it accepts a Shadowsocks configuration in JSON format. The format should match [the `ShadowsocksSessionConfig` defined in Outline Client](https://github.com/Jigsaw-Code/outline-client/blob/c29197957a6cb2ea84ba4d619d97f4d836f98094/src/www/app/tunnel.ts#L15-L21).

Related PRs: #108 , #111

### CODE REVIEW

File: outline/shadowsocks/client.go [hunk 1]
```diff
@@ -22,36 +27,70 @@ import (
...
 	"github.com/Jigsaw-Code/outline-go-tun2socks/outline/neterrors"
 	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
 	"github.com/Jigsaw-Code/outline-internal-sdk/transport/shadowsocks"
 	"github.com/Jigsaw-Code/outline-internal-sdk/transport/shadowsocks/client"
 	"github.com/eycorsican/go-tun2socks/common/log"
 )
 
-// Config represents a shadowsocks server configuration.
-// Exported via gobind.
-type Config struct {
```
Comment (fortuna) [inline outline/shadowsocks/client.go:34]: We need to keep this for now, for backward compatibility. But mark it as deprecated.

Breaking the binary api will break nthlink, and will prevent us from incrementally upgrading the client to the new api.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: outline/internal/utf8/parser.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// DecodeCodepointsToBytes parses a UTF-8 string as a raw byte array. That is
+// to say, each codepoint in the Unicode string will be treated as a single
+// byte (must be in range 0x00 ~ 0xff).
+//
+// If a codepoint falls out of the range, an error will be returned.
+func DecodeCodepointsToBytes(s string) ([]byte, error) {
+	runes := []rune(s)
+	rawBytes := make([]byte, len(runes))
+	for i, r := range runes {
+		if (r & 0xFF) != r {
```
[the remark this answers is not in the document]
Comment (jyyi1) [reply outline/internal/utf8/parser.go:29]: I feel it's not `json` specific, how about `DecodeUTF8CodepointsToRawBytes` (which is similar to [`base64.DecodeString`](https://pkg.go.dev/encoding/base64#Encoding.DecodeString)?

File: outline/shadowsocks/client.go [hunk 3]
```diff
@@ -22,36 +27,70 @@ import (
...
+
+// NewClientFromParameters creates a new Shadowsocks client from the parameters.
+//
+//   - host specifies the hostname/IP of the Shadowsocks server
+//   - port specifies the port number of the Shadowsocks server
+//   - cipher and password specifies the encryption strategy
+//   - prefix is an optional string, it specifies the salt prefix used in the
+//     beginning of the session. The should be a UTF-8 encoded string with each
+//     codepoint representing a single byte (0x00 ~ 0xFF) in the salt prefix.
+func NewClientFromParameters(host string, port int, cipher, password string, prefix string) (*Client, error) {
```
[the remark this answers is not in the document]
Comment (jyyi1) [reply outline/shadowsocks/client.go:68]: The difference is this function uses `prefix` as a string, but the legacy `NewClient` uses it as a `byte[]`. I'm trying to prevent the decode process happening from multiple places.

File: outline/shadowsocks/client.go (hunk 1 above)
Comment (jyyi1) [reply outline/shadowsocks/client.go:34]: It's moved to `config.go`.

File: outline/electron/main.go [hunk 4]
```diff
@@ -85,44 +91,23 @@ func main() {
...
+		} else {
+			client = c
+		}
+	} else {
+		if c, err := shadowsocks.NewClientFromJSON(jsonConfig); err != nil {
+			log.Errorf("Failed to construct Shadowsocks client: %v", err)
+			os.Exit(neterrors.IllegalConfiguration.Number())
+		} else {
+			client = c
 		}
```
[the remark this answers is not in the document]
Comment (jyyi1) [reply outline/electron/main.go:110]: I hope so too, but it seems go doesn't like that. (check the playground link above)

File: outline/electron/main.go [hunk 5]
```diff
@@ -85,44 +91,23 @@ func main() {
...
+	var client *shadowsocks.Client
+	if jsonConfig := *args.proxyConfig; len(jsonConfig) == 0 {
+		// legacy raw flags
+		if c, err := shadowsocks.NewClientFromParameters(*args.proxyHost, *args.proxyPort,
+			*args.proxyCipher, *args.proxyPassword, *args.proxyPrefix); err != nil {
+			log.Errorf("Failed to create Shadowsocks client: %v", err)
 			os.Exit(neterrors.IllegalConfiguration.Number())
+		} else {
+			client = c
+		}
```
[the remark this answers is not in the document]
Comment (fortuna) [reply outline/electron/main.go:103]: Use `=` instead of `:=`
The point is that the else is unnecessary and hurts readability.

File: outline/electron/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (fortuna) [reply outline/electron/main.go:110]: Same, use `=` instead to not create a new variable.

File: outline/internal/utf8/parser.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+package utf8
+
+import "fmt"
+
+// DecodeCodepointsToBytes parses a UTF-8 string as a raw byte array. That is
+// to say, each codepoint in the Unicode string will be treated as a single
+// byte (must be in range 0x00 ~ 0xff).
+//
+// If a codepoint falls out of the range, an error will be returned.
+func DecodeCodepointsToBytes(s string) ([]byte, error) {
```
Comment (fortuna) [inline outline/internal/utf8/parser.go:25]: Perhaps make it clear that the input string must be UTF8?
```suggestion
func DecodeCodepointsToBytes(utf8String string) ([]byte, error) {
```

File: outline/internal/utf8/parser.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fortuna) [reply outline/internal/utf8/parser.go:29]: `DecodeUTF8CodepointsToRawBytes` sounds good.

BTW, I found this article really helpful: https://go.dev/blog/strings

File: outline/electron/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jyyi1) [reply outline/electron/main.go]: I refactored it by adding a new private function `newShadowsocksClientFromArgs`, so both the `err` objects and `string`->`byte[]` decoding and can be handled in a cleaner way.

File: outline/shadowsocks/client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyyi1) [reply outline/shadowsocks/client.go]: Removed the new function, adding a private function to `main.go` instead.

**Review by fortuna — APPROVED** [review]
Looks great! Thanks for getting this done!

It would be nice to add a test case to exercise the prefix corner case.

File: outline/shadowsocks/config_test.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+		},
+		{
+			name:  "unprintable prefix",
+			input: `{"host":"192.0.2.1","port":12345,"method":"some-cipher","password":"abcd1234","prefix":"abc 123","prefix":"\u0000\u0080\u00ff"}`,
+			want: &configJSON{
+				Host:     "192.0.2.1",
+				Port:     12345,
+				Method:   "some-cipher",
+				Password: "abcd1234",
+				Prefix:   "\u0000\u0080\u00ff",
```
[the remark this answers is not in the document]
Comment (jyyi1) [reply outline/shadowsocks/config_test.go:69]: added `multi-byte utf-8 prefix`

**Review by fortuna — APPROVED** [review]
